### PR TITLE
Refactoring; use GenericCompositeFuture

### PR DESCRIPTION
**Pull request by adamdickmeiss**
Use Collections.singletonList for one arg. Use GenericCompositeFuture
instead of CompositeFuture. Many other minor changes.
Event `moduleUpdate` code removed as this is no longer in use.

### CODE REVIEW

File: okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java [hunk 1]
```diff
@@ -658,18 +654,23 @@ private void fireTimer(Tenant tenant, ModuleDescriptor md, RoutingEntry re, Stri
...
-    } else {
-      return Future.failedFuture(new OkapiError(ErrorType.USER,
-          "Unknown version of _tenantPermissions interface in use " + permIntVer + "."));
+    switch (permIntVer) {
+      case "1.0":
+        pl = new PermissionList(moduleTo, stripPermissionReplaces(mdTo.getPermissionSets()));
+        break;
+      case "1.1":
+        pl = new PermissionList(moduleTo, stripPermissionReplaces(
+            mdTo.getExpandedPermissionSets()));
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java:666]: inconsistent line break

File: okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java [hunk 2]
```diff
@@ -926,12 +927,8 @@ private static ModuleInstance getTenantInstanceForInterfacev2(
...
 
   Future<List<InstallJob>> installUpgradeGetList(String tenantId) {
-    return tenants.getNotFound(tenantId).compose(x -> jobs.get(tenantId).compose(list -> {
-      if (list == null) {
-        return Future.succeededFuture(new LinkedList<>());
-      }
-      return Future.succeededFuture(list);
-    }));
+    return tenants.getNotFound(tenantId).compose(x -> jobs.get(tenantId)
+        .map(list -> Objects.requireNonNullElseGet(list, LinkedList::new)));
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java:931]:
```suggestion
    return tenants.getNotFound(tenantId)
        .compose(x -> jobs.get(tenantId))
        .map(list -> Objects.requireNonNullElseGet(list, LinkedList::new));
```

File: okapi-core/src/main/java/org/folio/okapi/service/impl/DockerModuleHandle.java [hunk 3]
```diff
@@ -402,13 +399,11 @@ private int getExposedPort(JsonObject b) {
...
-      Map.Entry<String, Object> next = iterator.next();
+    for (Map.Entry<String, Object> next : exposedPorts) {
       String key = next.getKey();
       String port = key.split("/")[0];
       if (exposedPort == 0) {
-        exposedPort = Integer.valueOf(port);
+        exposedPort = Integer.parseInt(port);
       }
     }
     return exposedPort;
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/service/impl/DockerModuleHandle.java:409]: for (Map.Entry<String, Object> next : exposedPorts) {
      String key = next.getKey();
      String port = key.split("/")[0];
      return Integer.parseInt(port);
    }
    return 0;

File: okapi-core/src/main/java/org/folio/okapi/service/impl/DockerModuleHandle.java [hunk 4]
```diff
@@ -445,7 +440,7 @@ private int getExposedPort(JsonObject b) {
         .compose(
             x -> deleteContainer(),
             // if stopContainer fails with e run deleteContainer but return original failure e
-            e -> deleteContainer().onComplete(x -> Future.failedFuture(e)))
+            e -> deleteContainer())
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/service/impl/DockerModuleHandle.java:443]: The comment and the code no longer match.

File: okapi-core/src/main/java/org/folio/okapi/service/impl/ProcessModuleHandle.java [hunk 5]
```diff
@@ -186,17 +186,17 @@ private NuProcess launch(EnvEntry[] env, String [] command) {
...
-        } else {
-          return Future.failedFuture(messages.getMessage("11503", Integer.toString(port)));
-        }
-      });
-    }, fail -> Future.succeededFuture());
+    return c.connect(port, "localhost").compose(
+        socket -> socket.close()
+            .otherwiseEmpty().compose(x -> {
+              if (iter == 0) {
+                return Future.failedFuture(messages.getMessage("11503", Integer.toString(port)));
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/service/impl/ProcessModuleHandle.java:193]: A unit test is missing for this failure.

File: okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java (hunk 1 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java:666]: ??

File: okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java (hunk 2 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java:931]: ok

File: okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java [hunk 6]
```diff
@@ -51,20 +52,20 @@
...
   private static final Logger logger = OkapiLogger.get();
-  private ModuleManager moduleManager;
+  private final ModuleManager moduleManager;
   private ProxyService proxyService = null;
   private DiscoveryManager discoveryManager;
   private final TenantStore tenantStore;
   private LockedTypedMap1<Tenant> tenants = new LockedTypedMap1<>(Tenant.class);
-  private String mapName = "tenants";
-  private LockedTypedMap2<InstallJob> jobs = new LockedTypedMap2<>(InstallJob.class);
+  private static final String MAP_NAMAE = "tenants";
```
[the remark this answers is not in the document]
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java:60]: yep.

File: okapi-core/src/main/java/org/folio/okapi/service/impl/DockerModuleHandle.java (hunk 3 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/service/impl/DockerModuleHandle.java:409]: ok.

File: okapi-core/src/main/java/org/folio/okapi/service/impl/DockerModuleHandle.java (hunk 4 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/service/impl/DockerModuleHandle.java:443]: According to my editor the Failure.future value is lost, which was not the intention.

How about
```java
        e -> deleteContainer().eventually(x -> Future.failedFuture(e)))
```

File: okapi-core/src/main/java/org/folio/okapi/service/impl/ProcessModuleHandle.java (hunk 5 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/service/impl/ProcessModuleHandle.java:193]: done

File: okapi-core/src/main/java/org/folio/okapi/service/impl/DockerModuleHandle.java (hunk 4 above)
Comment (julianladisch) [reply okapi-core/src/main/java/org/folio/okapi/service/impl/DockerModuleHandle.java]: eventually is correct.

**Review by julianladisch — APPROVED** [review]
Thanks, looks good!

**Pull request merged**